DataFormats/Common/interface/getThinned_implementation.h

```diff
@@ -133,36 +203,235 @@ namespace edm {
+                                                foundContainers,
+                                                thinnedIndexes);
+          if (slimmed.thinnedAvailable()) {
+            thinnedAvailable = true;
+          } else if (slimmed.hasSlimmed()) {
+            assert(slimmedAssociation == nullptr);
+            std::tie(slimmedAssociation, slimmedIndexes) = slimmed.moveSlimmedAssociation();
+          }
+          for (unsigned k = 0; k < nKeys; ++k) {
+            if (foundContainers[k] == nullptr)
+              continue;
+            if (thinnedIndexes[k] == kThinningDoNotLookForThisIndex)
+              continue;
+            keys[k] = thinnedIndexes[k];
+          }
+        } else {
+          for (unsigned k = 0; k < nKeys; ++k) {
+            if (thinnedIndexes[k] == kThinningDoNotLookForThisIndex)
+              continue;
+            keys[k] = thinnedIndexes[k];
+            foundContainers[k] = thinnedCollection;
+          }
+        }
+      }
+      if (not thinnedAvailable and slimmedAssociation != nullptr) {
+        return SlimmedProducts(slimmedAssociation, std::move(slimmedIndexes));
+      }
+      return SlimmedProducts(thinnedAvailable);
+    }
+
+    // This function provides a common implementation of
+    // EDProductGetter::getThinnedProduct() for EventPrincipal,
+    // DataGetterHelper, and BareRootProductGetter.
+    //
+    // getThinnedProduct assumes getIt was already called and failed to find
+    // the product. The input key is the index of the desired element in the
+    // container identified by ProductID (which cannot be found).
+    // If the return value is not null, then the desired element was
+    // found in a thinned container. If the desired element is not
+    // found, then an optional without a value is returned.
+    template <typename F1, typename F2, typename F3>
+    std::optional<std::tuple<WrapperBase const*, unsigned int>> getThinnedProduct(
+        ProductID const& pid,
+        unsigned int key,
+        ThinnedAssociationsHelper const& thinnedAssociationsHelper,
+        F1 pidToBid,
+        F2 getThinnedAssociation,
+        F3 getByProductID) {
+      auto thinnedOrSlimmed =
+          getThinnedOnlyProduct(pid, key, thinnedAssociationsHelper, pidToBid, getThinnedAssociation, getByProductID);
+
+      if (thinnedOrSlimmed.hasThinned()) {
+        return thinnedOrSlimmed.thinnedProduct();
+      } else if (thinnedOrSlimmed.hasSlimmed()) {
+        auto [slimmedAssociation, slimmedIndex] = thinnedOrSlimmed.slimmedAssociation();
+        ProductID const& slimmedCollectionPID = slimmedAssociation->thinnedCollectionID();
+        WrapperBase const* slimmedCollection = getByProductID(slimmedCollectionPID);
+        if (slimmedCollection == nullptr) {
+          // Slimmed container is not found, try looking recursively in thinned containers
+          // which were made by selecting elements from this thinned container.
+          return getThinnedProduct(slimmedCollectionPID,
```

Comment:
I think this code is doing what you intended, but this text in the rules at the top of the PR doesn't seem to describe it precisely, "else if there is a slimmed collection, return element in it, or nullptr if the slimmed collection does not contain the element". Maybe this "Else it searches for the first slimmed collection working downward through the graph of related slimmed/thinned collections. If it finds one, then it recursively applies this algorithm to the slimmed collection as if it were the parent collection." For example, a slimmed collection is made. Then a slimmed collection and a thinned-only collection are made from the slimmed collection. The first slimmed collection is dropped and the thinned-only collection does not contain the element. But the second slimmed collection does contain the element and is present. null gets returned. I think this is the intent. In which case the code is correct, but the rule description is not.

Reply:
> For example, a slimmed collection is made. Then a slimmed collection and a thinned-only collection are made from the slimmed collection. The first slimmed collection is dropped and the thinned-only collection does not contain the element. But the second slimmed collection does contain the element and is present. null gets returned. I think this is the intent.

This is indeed the intent.

> I think this is the intent. In which case the code is correct, but the rule description is not.

Thanks, I see I had implicitly implied the "apply algorithm recursively on the slimmed collection as a parent". I've rephrased the PR description to make the recursion explicit.